test/test_evaluators.py

```diff
@@ -7,10 +7,93 @@
     PrecisionEvaluator, RecallEvaluator, TagWiseAccuracyEvaluator, TagWiseAveragePrecisionEvaluator, MeanAveragePrecisionNPointsEvaluator, BalancedAccuracyScoreEvaluator, \
     CocoMeanAveragePrecisionEvaluator, BleuScoreEvaluator, METEORScoreEvaluator, ROUGELScoreEvaluator, CIDErScoreEvaluator, SPICEScoreEvaluator, RocAucEvaluator, MeanIOUEvaluator, \
-    ForegroundIOUEvaluator, BoundaryMeanIOUEvaluator, BoundaryForegroundIOUEvaluator, L1ErrorEvaluator
+    ForegroundIOUEvaluator, BoundaryMeanIOUEvaluator, BoundaryForegroundIOUEvaluator, L1ErrorEvaluator, GroupWiseEvaluator
 from vision_evaluation.prediction_filters import TopKPredictionFilter, ThresholdPredictionFilter
 
 
+class TestGroupWiseEvaluator(unittest.TestCase):
+    def test_group_wise_classification_evaluator(self):
+
+        GROUND_TRUTHS = [{"targets": np.array([0, 0, 0, 0, 2, 2]),
+                          "groups": np.array([0, 0, 1, 1, 0, 0])},
+
+                         {"targets": np.array([0, 0, 0, 1, 1, 1]),
+                          "groups": np.array([0, 0, 0, 1, 1, 1])}]
+
+        PREDICTIONS = [np.array([[0.8, 0.1, 0.1],
+                                [0.8, 0.1, 0.1],
+                                [0.1, 0.8, 0.1],
+                                [0.1, 0.8, 0.1],
+                                [0.1, 0.1, 0.8],
+                                [0.1, 0.1, 0.8]]),
+
+                       np.array([[0.8, 0.1],
+                                [0.1, 0.8],
+                                [0.8, 0.1],
+                                [0.1, 0.8],
+                                [0.8, 0.1],
+                                [0.1, 0.8]])]
+
+        N_TARGET_CLASSES = [3, 2]
+        N_GROUP_CLASSES = [2, 2]
+
+        ACCURACY = [[{0: 1,
+                    1: 0},
+                    {0: 1,
+                    1: 1}],
+
+                    [{0: 0.666,
+                      1: 0.666},
+                    {0: 1,
+                     1: 1}]]
+
+        for k_idx, top_k in enumerate([1, 5]):
+
+            for i, (ground_truths, predictions) in enumerate(zip(GROUND_TRUTHS, PREDICTIONS)):
+
+                eval_fn = TopKAccuracyEvaluator(top_k)
+                eval = GroupWiseEvaluator(eval_fn)
+
+                eval.add_predictions(predictions, ground_truths)
+                group_wise_top_k_acc = eval.get_report()['group_wise_metrics']
+
+                if top_k >= N_TARGET_CLASSES[i]:
+                    import warnings
+                    warnings.warn(f"k{[top_k]} > no. of target classes{[N_TARGET_CLASSES]} will result in a perfect score and is therefore meaningless.")
+
+                for g in range(N_GROUP_CLASSES[i]):
+
+                    if g in group_wise_top_k_acc:
+                        print(f"top_k: {top_k} | sample no: {i+1} | group: {g}")
```

Comment:
dont use print for logging
dont log in tests

Reply:
updated